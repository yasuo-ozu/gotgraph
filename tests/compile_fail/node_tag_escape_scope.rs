use gotgraph::prelude::*;

fn main() {
    let mut graph: VecGraph<i32, &str> = VecGraph::default();
    
    // Try to move node tags outside of scope - should fail
    let escaped_node = graph.scope_mut(|mut ctx| {
        let n0 = ctx.add_node(42);
        n0 // ERROR: node tag cannot escape the scope
    });
    
    // Try to use the escaped node tag - should fail
    println!("{:?}", escaped_node);
}