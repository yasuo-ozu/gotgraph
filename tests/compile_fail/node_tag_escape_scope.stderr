error: lifetime may not live long enough
 --> tests/compile_fail/node_tag_escape_scope.rs:9:9
  |
7 |     let escaped_node = graph.scope_mut(|mut ctx| {
  |                                         -------- return type of closure is NodeTag<'2, NodeIx>
  |                                         |
  |                                         has type `gotgraph::graph::Context<'1, &mut gotgraph::vec_graph::VecGraph<i32, &str>>`
8 |         let n0 = ctx.add_node(42);
9 |         n0 // ERROR: node tag cannot escape the scope
  |         ^^ returning this value requires that `'1` must outlive `'2`
  |
  = note: requirement occurs because of the type `NodeTag<'_, NodeIx>`, which makes the generic argument `'_` invariant
  = note: the struct `NodeTag<'scope, I>` is invariant over the parameter `'scope`
  = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
